Add compile_library & compile_tx_script helper functions

partylikeits1983 opened the pull request:
## Overview
This PR adds two helper functions to a newly created `client_utils` module.

1. `compile_library` - creates a miden assembly library from the provided account code and library path.
2. `compile_tx_script` -  creates a transaction script based on the provided transaction script code and library.

These two helper functions compliment each other since you'd first call `compile_library` then using the returned library, pass it into `compile_tx_script`. 

## Motivation
This PR was created to implement this: https://github.com/0xMiden/miden-client/issues/905#issuecomment-2920344436

The inspiration for this PR was from the `miden-client-tools` crate I wrote to minimize the lines of code required to get started writing and deploying applications on miden: https://crates.io/crates/miden-client-tools

## Comments
This PR is sort of a WIP. Please let me know if the `client_utils` module should be renamed, moved, or completely changed.

In this PR I added a test and also tested the functions in a different project to feel out how these functions work in practice. 

I think the `client_utils` module could also serve as a springboard for adding other utility functions, such as `Falcon512` signing utilities, and other various helper functions that minimize lines of code developers need to add to their projects.

partylikeits1983 commented on crates/rust-client/src/client_utils/mod.rs:
```diff
@@ -0,0 +1,54 @@
+/// # Returns
+///
+/// Returns a `TransactionScript` if successfully created, or an error.
+pub fn compile_tx_script(
+    script_code: String,
+    library: Library,
+) -> Result<TransactionScript, Box<dyn std::error::Error>> {
+    let assembler = TransactionKernel::assembler().with_library(library)?;
+
+    Ok(TransactionScript::compile(script_code, [], assembler)?)
+}
```
Should we make it an optional parameter? Sometimes you don't need a library.

tomyrd replied:
Yeah, I agree

igamigo commented on Cargo.toml:
```diff
@@ -24,6 +24,7 @@ miden-lib = { version = "0.9" , default-features = false }
 miden-objects = { version = "0.9" , default-features = false }
 miden-tx = { version = "0.9" , default-features = false, features = ["async"] }
 miden-testing = { version = "0.9" , default-features = false, features = ["async"] }
+miden-assembly = { version = "0.14.0" }
```
I'm not entirely sure we need to add `miden-assembly`. I could be wrong but I think the structs you are importing are re-exported from `miden-objects`

partylikeits1983 commented on crates/rust-client/src/client_utils/mod.rs:
```diff
@@ -0,0 +1,54 @@
+
+    let library = assembler.assemble_library([module])?;
+    Ok(library)
+}
```
Added it to the `Client` struct.

partylikeits1983 commented:
Removed the `client_utils` module, and added the `compile_library` function to the `Client` struct. Also updated the `compile_tx_script` function to take an optional `Library` parameter.

tomyrd commented on crates/rust-client/src/transaction/mod.rs:
```diff
@@ -813,16 +813,49 @@ impl Client {
     }
 
-    /// Compiles the provided transaction script source and inputs into a [`TransactionScript`].
+    /// Compiles the provided account code, library path into a [`Library`]
+    pub fn compile_library(
+        &self,
+        account_code: &str,
+        library_path: &str,
+    ) -> Result<Library, ClientError> {
```
nit: A compiled library can contain other code that isn't account, am I right? (Although account code would be the most common use). If so, I would change the docs rename the `account_code` to something like `library_code`.

partylikeits1983 replied:
Yes, good point. Changed it to `library_code` but kept the name `account_code` in the test to show that you can pass an account's code to the `compile_library` function.

igamigo commented on crates/rust-client/src/transaction/mod.rs:
```diff
@@ -813,16 +813,49 @@ impl Client {
+    pub fn compile_library(
+        &self,
+        library_code: &str,
+        library_path: &str,
```
Should this parameter be `LibraryPath` instead? If you think it's easier to abstract this away, we could keep it a `&str` but it sounds better to leverage the types that are there to improve the API. Also, would be nice to explain briefly what these parameters are for as part of the doc comments, since it will be a public function.

igamigo commented on crates/rust-client/src/transaction/mod.rs:
```diff
@@ -813,16 +813,49 @@ impl Client {
     pub fn compile_tx_script<T>(
         &self,
         inputs: T,
+        library: Option<Library>,
```
Does it make sense to make this `Vec<Library>`? If the user wants to compile a transaction against multiple libraries, it looks like there would be no way to do it right now.

igamigo reviewed (commented):
Left some comments but overall looks good!

bobbinth commented on crates/rust-client/src/transaction/mod.rs:
```diff
@@ -813,16 +813,49 @@ impl Client {
 
-    /// Compiles the provided transaction script source and inputs into a [`TransactionScript`].
+    /// Compiles the provided library code and library path into a [`Library`]
+    pub fn compile_library(
+        &self,
+        library_code: &str,
+        library_path: &str,
+    ) -> Result<Library, ClientError> {
```
Why do we need to pass in `&self` into this function? AFAICT, it is not being used in the body. This also means that this could probably be a "free" utility function rather than a method on the `Client` struct.

bobbinth replied:
My understanding is that the primary motivation behind these functions is to simplify transaction script building. Instead of putting these on the `Client` or having free functions, it may make sense to have something like a `ScriptBuilder`. It could look something like this:

```rust
impl ScriptBuilder {

    pub fn new(in_debug_mode: bool) -> Self {
        ...
    }

    pub fn add_library(
        &mut self,
        library_code: impl AsRef<str>,
        library_path: impl TryInto<LibraryPath>,
    ) -> Result<(), SomeError> {
        ...
    }

    pub fn compile_tx_script(
        &self,
        program: impl AsRef<str>,
        inputs: impl IntoIterator<Item = (Word, Vec<Felt>)>,
    ) -> Result<TransactionScript, SomeError> {
        ...
    }
}
```
We can even add a method on the client to get the builder:
```rust
impl Client {
    pub fn script_builder(&self) -> ScriptBuilder {
        ScriptBuilder::new(self.in_debug_mode)
    }
}
```

We then will be able to do something like:
```rust
let mut builder = client.script_builder();
builder.add_library(lib_code, lib_path)?;
let script = builder.compile_tx_script(script_code, [])?;
```
In the future, we could extend the `ScriptBuilder` to allow loading libraries from files and directories. We can also make it more general to allow compiling note scripts as well.

igamigo commented:
Closing this in favor of #1022
